### PR TITLE
WFLY-9710 Add Undertow specific buffer pool

### PR DESCRIPTION
**Pull request by stuartwdouglas**
https://issues.jboss.org/browse/WFLY-9710

### CODE REVIEW

File: undertow/src/main/java/org/wildfly/extension/undertow/ByteBufferPoolDefinition.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+    private static class BufferPoolAdd extends AbstractAddStepHandler {
+
+        private BufferPoolAdd() {
+            super(ByteBufferPoolDefinition.ATTRIBUTES);
+        }
+
+        @Override
+        protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+            final PathAddress address = PathAddress.pathAddress(operation.get(OP_ADDR));
```
Comment (bstansberry) [inline undertow/src/main/java/org/wildfly/extension/undertow/ByteBufferPoolDefinition.java]: final PathAddress address = context.getCurrentAddress();

File: undertow/src/main/java/org/wildfly/extension/undertow/ByteBufferPoolDefinition.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            final ModelNode bufferSizeModel = BUFFER_SIZE.resolveModelAttribute(context, model);
+            final ModelNode maxPoolSizeModel = MAX_POOL_SIZE.resolveModelAttribute(context, model);
+            final ModelNode directModel = DIRECT.resolveModelAttribute(context, model);
+            final int threadLocalCacheSize = THREAD_LOCAL_CACHE_SIZE.resolveModelAttribute(context, model).asInt();
+            final int leakDetectionPercent = LEAK_DETECTION_PERCENT.resolveModelAttribute(context, model).asInt();
+
+            final int bufferSize = bufferSizeModel.isDefined() ? bufferSizeModel.asInt() : defaultBufferSize;
```
Comment (bstansberry) [inline undertow/src/main/java/org/wildfly/extension/undertow/ByteBufferPoolDefinition.java]: no need to change this but just fyi this and the others can be

final int bufferSize = bufferSizeModel.asInt(defaultBufferSize);

File: undertow/src/main/java/org/wildfly/extension/undertow/ListenerResourceDefinition.java [hunk 3]
```diff
@@ -91,13 +90,16 @@
...
             .build();
+
     protected static final SimpleAttributeDefinition BUFFER_POOL = new SimpleAttributeDefinitionBuilder(Constants.BUFFER_POOL, ModelType.STRING)
             .setRequired(false)
             .setFlags(AttributeAccess.Flag.RESTART_ALL_SERVICES)
             .setValidator(new StringLengthValidator(1))
             .setDefaultValue(new ModelNode("default"))
-            .setCapabilityReference(REF_BUFFER_POOL, LISTENER_CAPABILITY)
+            .setCapabilityReference(Capabilities.CAPABILITY_BYTE_BUFFER_POOL, LISTENER_CAPABILITY)
+            .setAlternatives(Constants.BYTE_BUFFER_POOL)
```
Comment (bstansberry) [inline undertow/src/main/java/org/wildfly/extension/undertow/ListenerResourceDefinition.java]: Is this left over from an earlier draft?

**Review by bstansberry — CHANGES_REQUESTED** [review]

**Comment by kabir** [comment]
Retest this please

**Comment by kabir** [comment]
Retest this please

**Comment by kabir** [comment]
@stuartwdouglas merge conflicts

**Comment by stuartwdouglas** [comment]
Retest this please

**Comment by stuartwdouglas** [comment]
Retest this please

**Comment by stuartwdouglas** [comment]
Retest this please

**Comment by stuartwdouglas** [comment]
This just implements the pools themselves, it does not change the config to make them default, as there are compatibility issues with older servers in mixed domain. I am not sure if we should just:
1) Leave this PR as is for now, and have the option of Undertow pools but XNIO still the default
2) Change the defaults, and potentially cause issues for mixed domain scenarios
3) Attempt to transform the Undertow pools to XNIO ones for older servers
4) Something else??

**Comment by kabir** [comment]
@stuartwdouglas I don't think we should magically try to change the undertow pools into xnio ones for older servers.

I've not looked at the PR in any detail, mainly at your list of options, but I think we should keep the option to have xnio pools. If it is desired, the undertow ones should be made default. For transformers/mixed-domain we should reject the usage of the undertow pools. Typically the way this works is that people have a 'current' profile, and if they also need to support older HCs where new things in the 'current' profile cannot be transformed, they create an 'old' profile containing stuff that can be understood on the older HCs. This is what the DomainAdjusters in the mixed-domain testsuite do (e.g. for transformation to 6.x they remove the undertow subsystem completely and add the old web subsystem).

**Comment by stuartwdouglas** [comment]
At the moment this PR just adds the option of using undertow pools, the XNIO ones stay as the default.

**Review by jstourac — APPROVED** [review]

**Comment by kabir** [comment]
@stuartwdouglas once this is rebased (and community docs added, if it makes sense) it should be good to go

**Comment by kabir** [comment]
@stuartwdouglas ^^

**Review by kabir — CHANGES_REQUESTED** [review]
@stuartwdouglas For completeness we should have test that the reject transformation works for both 7.1.0 and 7.0.0. 

I realise this conflicts with https://github.com/wildfly/wildfly/pull/10996#pullrequestreview-116526428 so perhaps a followup PR containing these two PRs and an extra commit to add the tests for both could work?

Also it is missing community docs.

**Pull request merged**